snuba/request_schema.py

```diff
@@ -0,0 +1,132 @@
+from __future__ import annotations
+
+import copy
+import itertools
+import jsonschema
+
+from collections import ChainMap
+from dataclasses import dataclass
+from typing import Any, Mapping
+
+from snuba.query.extensions import QueryExtension
+from snuba.query.query import Query
+from snuba.schemas import Schema, GENERIC_QUERY_SCHEMA
+
+
+@dataclass(frozen=True)
+class Request:
+    query: Query
+    extensions: Mapping[str, Mapping[str, Any]]
+
+    @property
+    def body(self):
+        return ChainMap(self.query.get_body(), *self.extensions.values())
+
+
+class RequestSchema:
+    def __init__(self, query_schema: Schema, extensions_schemas: Mapping[str, Schema]):
+        self.__query_schema = query_schema
+        self.__extension_schemas = extensions_schemas
+
+        self.__composite_schema = {
+            'type': 'object',
+            'properties': {},
+            'required': [],
+            'definitions': {},
+            'additionalProperties': False,
+        }
+
+        for schema in itertools.chain([self.__query_schema], self.__extension_schemas.values()):
+            assert schema['type'] == 'object', 'subschema must be object'
+            assert schema['additionalProperties'] is False, 'subschema must not allow additional properties'
+            self.__composite_schema['required'].extend(schema.get('required', []))
+
+            for property_name, property_schema in schema['properties'].items():
+                assert property_name not in self.__composite_schema['properties'], 'subschema cannot redefine property'
+                self.__composite_schema['properties'][property_name] = property_schema
+
+            for definition_name, definition_schema in schema.get('definitions', {}).items():
+                assert definition_name not in self.__composite_schema['definitions'], 'subschema cannot redefine definition'
+                self.__composite_schema['definitions'][definition_name] = definition_schema
+
+        self.__composite_schema['required'] = set(self.__composite_schema['required'])
+
+    @classmethod
+    def build_with_extensions(cls, extensions: Mapping[str, QueryExtension]) -> RequestSchema:
```

Comment:
Ok I lied, there was a functional change: this factory method is new and it is the whole reason to move this class to a separate file, but it is the only one.

Reply:
I like this.